File: src/main/java/com/herbertgao/telegram/bot/gaokao/business/service/InlineQueryService.java
package com.herbertgao.telegram.bot.gaokao.business.service;

import com.herbertgao.telegram.bot.gaokao.business.common.util.GaokaoBotUtil;
import com.herbertgao.telegram.bot.gaokao.business.common.util.IdUtil;
import com.herbertgao.telegram.bot.gaokao.database.domain.ExamDate;
import com.herbertgao.telegram.bot.gaokao.database.domain.UserTemplate;
import com.herbertgao.telegram.bot.gaokao.database.service.ExamDateService;
import com.herbertgao.telegram.bot.gaokao.database.service.UserTemplateService;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.telegram.telegrambots.meta.api.methods.AnswerInlineQuery;
import org.telegram.telegrambots.meta.api.objects.inlinequery.InlineQuery;
import org.telegram.telegrambots.meta.api.objects.inlinequery.inputmessagecontent.InputTextMessageContent;
import org.telegram.telegrambots.meta.api.objects.inlinequery.result.InlineQueryResult;
import org.telegram.telegrambots.meta.api.objects.inlinequery.result.InlineQueryResultArticle;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Inline消息服务
 *
 * @author HerbertGao
 * @date 2021-12-10
 */
@Slf4j
@Service
public class InlineQueryService {

    @Autowired
    private ExamDateService examDateService;
    @Autowired
    private UserTemplateService userTemplateService;

    public AnswerInlineQuery answerInlineQuery(InlineQuery inlineQuery) {

        List<InlineQueryResult> resultList = new ArrayList<>();

        Long userId = inlineQuery.getFrom().getId();
        String query = inlineQuery.getQuery();
        LocalDateTime now = LocalDateTime.now().withNano(0);
        List<ExamDate> examList = new ArrayList<>();

        if (StringUtils.isNotBlank(query)) {
            if (StringUtils.isNumeric(query)) {
                Integer year = Integer.valueOf(query);
                examList = examDateService.getExamByYear(year);
            }
        } else {
            examList = examDateService.getExamList(now, false);
        }

        UserTemplate defaultTemplate = userTemplateService.getDefaultTemplate();
        List<UserTemplate> templateList = userTemplateService.getUserTemplateListByUserId(userId);

        for (ExamDate exam : examList) {

            String examDesc = exam.getShortDesc();

            String defaultTitle = "查看" + examDesc + "倒计时";
            String defaultMessage = GaokaoBotUtil.getCountDownString(exam, now, defaultTemplate.getTemplateContent());
            InlineQueryResultArticle r = InlineQueryResultArticle.builder()
                    .id(IdUtil.getId().toString())
                    .title(defaultTitle)
                    .inputMessageContent(InputTextMessageContent.builder()
                            .messageText(defaultMessage)
                            .build())
                    .build();
            resultList.add(r);

            templateList.forEach(template -> {
                String title = "查看" + examDesc + "倒计时";
                if (StringUtils.isNotBlank(template.getTemplateName())) {
                    title += " (" + template.getTemplateName() + ")";
                }
                String message = GaokaoBotUtil.getCountDownString(exam, now, template.getTemplateContent());
                InlineQueryResultArticle ru = InlineQueryResultArticle.builder()
                        .id(IdUtil.getId().toString())
                        .title(title)
                        .inputMessageContent(InputTextMessageContent.builder()
                                .messageText(message)
                                .build())
                        .build();
                resultList.add(ru);
            });
        }

        String inlineQueryId = inlineQuery.getId();

        AnswerInlineQuery answerInlineQuery = AnswerInlineQuery.builder()
                .inlineQueryId(inlineQueryId)
                .results(resultList)
                .cacheTime(1)
                .build();
        return answerInlineQuery;

    }

}

File: internal/service/inline_query_service.go
package service

import (
	"fmt"
	"strconv"

	"github.com/herbertgao/gaokao_bot/internal/model"
	"github.com/herbertgao/gaokao_bot/internal/util"
	"github.com/herbertgao/gaokao_bot/pkg/constant"
	"github.com/mymmrac/telego"
	"github.com/sirupsen/logrus"
)

// InlineQueryService 内联查询服务
type InlineQueryService struct {
	examDateService     *ExamDateService
	userTemplateService *UserTemplateService
	logger              *logrus.Logger
}

// NewInlineQueryService 创建内联查询服务
func NewInlineQueryService(
	examDateService *ExamDateService,
	userTemplateService *UserTemplateService,
	logger *logrus.Logger,
) *InlineQueryService {
	return &InlineQueryService{
		examDateService:     examDateService,
		userTemplateService: userTemplateService,
		logger:              logger,
	}
}

// GetInlineQueryResults 获取内联查询结果
func (s *InlineQueryService) GetInlineQueryResults(query *telego.InlineQuery) []telego.InlineQueryResult {
	now := util.NowBJT()

	var examList []model.ExamDate
	var err error

	// 解析年份
	if query.Query != "" {
		if y, err := strconv.Atoi(query.Query); err == nil && y >= constant.MinExamYear && y <= constant.MaxExamYear {
			examList, err = s.examDateService.GetExamByYear(y)
			if err != nil {
				s.logger.Errorf("按年份 %d 查询考试失败: %v", y, err)
				return []telego.InlineQueryResult{}
			}
		}
	} else {
		// 没有参数时，获取当前时间范围内的所有考试
		examList, err = s.examDateService.GetExamsInRange(now)
		if err != nil {
			s.logger.Errorf("查询时间范围内的考试失败: %v", err)
			return []telego.InlineQueryResult{}
		}
	}

	// 如果没有找到任何考试
	if len(examList) == 0 {
		return []telego.InlineQueryResult{}
	}

	// 获取默认模板
	defaultTemplate, err := s.userTemplateService.GetDefaultTemplate()
	if err != nil {
		s.logger.Errorf("获取默认模板失败: %v", err)
		return []telego.InlineQueryResult{}
	}

	// 获取用户自定义模板
	var userTemplates []model.UserTemplate
	if query.From.ID != 0 {
		userTemplates, _ = s.userTemplateService.GetByUserID(query.From.ID)
	}

	results := []telego.InlineQueryResult{}

	// 为每个考试生成inline结果
	for idx, exam := range examList {
		examDesc := exam.ShortDesc

		// 默认模板结果
		if defaultTemplate != nil {
			defaultTitle := fmt.Sprintf("查看%s倒计时", examDesc)
			defaultMessage := util.GetCountDownString(&exam, defaultTemplate.TemplateContent, now)
			result := &telego.InlineQueryResultArticle{
				Type:  telego.ResultTypeArticle,
				ID:    fmt.Sprintf("default_%d", idx),
				Title: defaultTitle,
				InputMessageContent: &telego.InputTextMessageContent{
					MessageText: defaultMessage,
				},
			}
			results = append(results, result)
		}

		// 用户自定义模板结果
		for tidx, template := range userTemplates {
			title := fmt.Sprintf("查看%s倒计时", examDesc)
			if template.TemplateName != "" {
				title = fmt.Sprintf("%s (%s)", title, template.TemplateName)
			}
			message := util.GetCountDownString(&exam, template.TemplateContent, now)
			result := &telego.InlineQueryResultArticle{
				Type:  telego.ResultTypeArticle,
				ID:    fmt.Sprintf("user_%d_%d", idx, tidx),
				Title: title,
				InputMessageContent: &telego.InputTextMessageContent{
					MessageText: message,
				},
			}
			results = append(results, result)
		}
	}

	return results
}
